spec.md

```diff
@@ -2734,6 +2790,257 @@ message NodeExpandVolumeResponse {
+
+This operation MUST be idempotent.
+If a group snapshot corresponding to the specified `group_snapshot_id` does not exist or the artifacts associated with the group snapshot do not exist anymore, the Plugin MUST reply `0 OK`.
+
+```protobuf
+message DeleteVolumeGroupSnapshotRequest {
+  option (alpha_message) = true;
+
+  // The ID of the group snapshot to be deleted.
+  // This field is REQUIRED.
+  string group_snapshot_id = 1;
+
+  // A list of snapshot IDs that are part of this group snapshot.
+  // If SP does not need to rely on this field to delete the snapshots
+  // in the group, this field MAY be ignored.
+  // Some SPs require this list to delete the snapshots in the group.
+  // If SP needs to use this field to delete the snapshots in the
+  // group, it MAY report an error when there is a mismatch.
+  // This field is REQUIRED.
+  repeated string snapshot_ids = 2;
+
+  // Secrets required by plugin to complete group snapshot deletion
+  // request.
+  // This field is OPTIONAL. Refer to the `Secrets Requirements`
+  // section on how to use this field.
+  // The secrets provided in this field SHOULD be the same for
+  // all group snapshot operations on the same group snapshot.
+  // The secrets provided in the field SHOULD be passed to both
+  // the group snapshot and the individual snapshot members if needed.
+  map<string, string> secrets = 3 [(csi_secret) = true];
+}
+
+message DeleteVolumeGroupSnapshotResponse {
+  // Intentionally empty.
+}
+```
+
+##### DeleteVolumeGroupSnapshot Errors
+
+If the plugin is unable to complete the DeleteVolumeGroupSnapshot call successfully, it MUST return a non-ok gRPC code in the gRPC status.
+If the conditions defined below are encountered, the plugin MUST return the specified gRPC error code.
+The CO MUST implement the specified error recovery behavior when it encounters the gRPC error code.
+
+| Condition | gRPC Code | Description | Recovery Behavior |
+|-----------|-----------|-------------|-------------------|
+| Volume group snapshot in use | 9 FAILED_PRECONDITION | Indicates that the volume group snapshot corresponding to the specified `group_snapshot_id` could not be deleted because it is in use by another resource. | Caller SHOULD ensure that there are no other resources using the volume group snapshot, and then retry with exponential back off. |
+
+#### `GetVolumeGroupSnapshot`
+
+**ALPHA FEATURE**
+
+This optional RPC MAY be called by the CO to fetch current information about a volume group snapshot.
+
+A Controller Plugin MUST implement this `GetVolumeGroupSnapshot` RPC call if it has `CREATE_DELETE_GET_VOLUME_GROUP_SNAPSHOT` capability.
+
+`GetVolumeGroupSnapshotResponse` should contain current information of a volume group snapshot if it exists.
+If the volume group snapshot does not exist any more, `GetVolumeGroupSnapshot` should return gRPC error code `NOT_FOUND`.
+
+```protobuf
+message GetVolumeGroupSnapshotRequest {
```

Comment:
it's interesting that the "Get" call doesn't have a provision for possibly requiring the individual IDs like the "Delete" call does. that seems dissonant and it's unclear to me how or why some systems would require the individual IDs for a Delete but not for a Get - can someone please provide an illustrating use case?

Reply:
Maybe we should still keep GetVolumeGroupSnapshot as a separate optional capability. @bswartz and I discussed about this before. At that time, we thought SP should be able to implement Get if they can implement Create/Delete.
@jdef made a good point here.  If SP does not track individual snapshots after the group snapshot is created, then it probably will have problems with both Get and Delete if individual snapshot ids are not provided as an input. Thoughts?

Reply:
I think @jdef is right about this. I can't recall thinking about whether Get did or did not need a list, and I probably just overlooked this. Based on the imaginary CSI driver I have in mind that tracks individual snapshots and has not object representing the group, you would need the list of snapshot IDs to return any kind of aggregated status. Then again, in the driver I have in mind, snapshots are instantly ready to use after being cut, so the mere existing of a group snapshot ID tells you that it's ready to use.

I'm not opposed to requiring the client to provide a list of snapshot ID. It's purely a CO-side burden, and I don't think it will be hard to code the sidecar to have that information available.

Reply:
Updated: https://github.com/container-storage-interface/spec/pull/519/commits/4ced11bfe8df7994223d3efb60f52eb65f18ddac